keps/sig-api-machinery/4933-cacher-label-index/README.md

```diff
@@ -0,0 +1,880 @@
+
+Each production k8s cluster in our company may run thousands of spark or flink jobs in parallel (the number of jobs may be roughly the same size as the number of nodes, or even higher). Each spark/flink job will start a master pod to create/list/watch/delete worker pods, which is Spark's `Driver` pod and Flink's `JobManager` pod, respectively. We found that this will cause a lot of pressure on the apiserver. In extreme cases, the number of pod add events sent by the apiserver to the scheduler per second will not keep up with the speed of pod creation per second, resulting in the scheduling speed not keeping up with the speed of pod creation, causing resource waste.
+
+Because a Spark or a Flink job only lists and watches the pod change events of itself by applicationId label, and the number of list and watch requests from these jobs is also comparable to the number of that from daemonsets in a considerable cluster. If we can have a label index as a cache index and trigger index, we can improve apiserver performance for running Spark and Flink jobs a lot.
+
+Although [PR#115918](https://github.com/kubernetes/kubernetes/pull/115918) has added namespace indexes to the apiserver cacher, this is not so useful in scenarios where a large number of jobs are running in a small number of namespaces. Most events are still sent to each cacheWatcher through the cacher first and then filtered, causing high CPU usage on the apiserver and delays in event delivery.
+
+### Background 1
+
+Each job of the big data computing engine spark/flink will start a master pod to create/list/watch/delete worker pods, namely the Driver pod of spark and the JobManager pod of flink.
+spark:
+- list: https://github.com/apache/spark/blob/branch-3.5/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsPollingSnapshotSource.scala#L77-L92
+- watch: https://github.com/apache/spark/blob/branch-3.5/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsWatchSnapshotSource.scala#L63-L67
+
+flink:
+- list: https://github.com/apache/flink/blob/master/flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClient.java#L207-L215
+- watch: https://github.com/apache/flink/blob/master/flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClient.java#L251-L257
+
+### Goals
+
+- Improve the efficiency of processing pod events of watch requests in apiserver when running Spark and Flink jobs.
+- Improve the processing efficiency of list requests with label selectors by adding label indexes.
+- Reduce the CPU usage of apiserver when running Spark and Flink jobs.
+- Indirectly improve cluster scheduling performance when cluster load is high when running Spark and Flink jobs.
```

Comment:
Why specifically Spark and Flink?

Reply:
You are right. I have add "The same improvements may apply to any other workloads as well if they also list and watch resources by labels." below to clarify this.